src/ipc-wrapper/server.js

```diff
@@ -0,0 +1,55 @@
+// @ts-check
+import { createServer } from 'rpc-reflector'
+import { SubChannel } from './sub-channel.js'
+
+/**
+ * @param {import('../mapeo-manager.js').MapeoManager} manager
+ * @param {import('rpc-reflector/server.js').MessagePortLike} messagePort
+ */
+export function createMapeoServer(manager, messagePort) {
+  // TODO: LRU? project.close() after time without use?
+  /** @type {Map<string, { close: () => void, project: import('../mapeo-project.js').MapeoProject }>}*/
+  const existingProjectServers = new Map()
+
+  const managerChannel = new SubChannel(messagePort, '@@manager')
+
+  const managerServer = createServer(manager, managerChannel)
+
+  messagePort.on('message', handleMessage)
+
+  return {
+    close() {
+      messagePort.off('message', handleMessage)
+
+      for (const [id, server] of existingProjectServers.entries()) {
+        server.close()
+        existingProjectServers.delete(id)
+      }
+
+      managerServer.close()
+    },
+  }
+
+  /**
+   * @param {any} payload
+   */
+  async function handleMessage(payload) {
+    const id = payload?.id
+
+    if (typeof id !== 'string' || id === '@@manager') return
+
+    if (existingProjectServers.has(id)) return
+
+    const projectChannel = new SubChannel(messagePort, id)
+
+    const project = await manager.getProject(
```

Comment:
Here is where your race condition is (always look for `await` in your code flow to find them). At this stage a second message for this project could come during the `await`.

One way you can solve this: add queuing to the SubChannel implementation (I think Tomás has done this for the implementation in the mapeo mobile new example), and only start this channel after you have done `createServer()` below. You will also need something like `existingSubChannels` which is set before this await, and check for that at the start of this `handleMessage` (so you won't get here twice - the queued messages in SubChannel will handle that)

Reply:
ahhh makes sense, thanks for explaining!

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/261/commits/9f6a7a5749f1b4c47ec4a1ffc44a18d29447bcc9